jack-core/src/com/rapleaf/jack/transaction/TransactorImpl.java

```diff
@@ -96,7 +96,14 @@ DbMetrics getDbMetrics() {
     } catch (Exception e) {
       LOG.error("SQL execution failure", e);
       if (asTransaction) {
-        connection.rollback();
+        try {
+          connection.rollback();
```

Comment:
Does it make sense to put this in a `try-finally` block? And remove this block below since the connection was returned?
``` java
if (asTransaction) {
  connection.rollback();
}
```

Reply:
Do you mean remove the `catch (Throwable t)` block below and add a `finally` block? I think it makes sense to do that, but there's this big block comment above that explicitly mentions why we don't do that. I sort of disagree with that comment though, so I'm hoping @sherifnada can weigh in as he added that comment and logic.

Reply:
I think josh is saying we should always return the connection, regardless of how the rollback went, so it makes sense to just put that particular piece inside a `finally` block, once, and put the rollback inside a `try` block... that way we have `dbManager.returnConnection(connection);` written only once here.

I agree with that idea.

Reply:
I think this should be resolved now with the recent changes.

Reply:
Hey guys, just catching up after coming back from a trip to China. It looks like this was merged already -- @briancecker did you feel that comment was inaccurate? or that it's the wrong way of fixing that problem?

Reply:
We believe it was the wrong way to fix the problem because it resulted in the `DBPoolManager` never letting go of stale connections, in certain situations, which killed a number of our apps as they could not get any new connections.

Reply:
As Henry mentioned, we didn't so much feel it was inaccurate, but the way of solving the issue lead to another issue. To preserve the behavior you added, avoid committing half-finished transactions, and avoid leaking connections, I added a new method for invalidating a connection instead of returning it. This method (`DbPoolManager#invalidateConnection`) removes the connection from the pool's tracked objects and closes it without committing.